src/legacy/shrink.rs

```diff
@@ -0,0 +1,388 @@
+        || codetab[code as usize].prefix_code == Some(code)
+    {
+        // Reject invalid codes. Self-referential codes may exist in
+        // the table but cannot be used.
+        return Err(io::Error::new(io::ErrorKind::InvalidData, "invalid code"));
+    }
+
+    if codetab[code as usize].len != UNKNOWN_LEN {
+        // Output string with known length (the common case).
+        let ct = &codetab[code as usize];
+        for i in ct.last_dst_pos..ct.last_dst_pos + ct.len as usize {
+            dst.push_back(dst[i]);
+        }
+        return Ok(dst[ct.last_dst_pos]);
+    }
+
+    // Output a string of unknown length. This happens when the prefix
+    // was invalid (due to partial clearing) when the code was inserted into
+    // the table. The prefix can then become valid when it's added to the
+    // table at a later point.
+    let prefix_code = codetab[code as usize].prefix_code.unwrap();
+    if cfg!(debug_assertions) {
+        let tab_entry = codetab[code as usize];
+        assert!(tab_entry.len == UNKNOWN_LEN);
+        assert!(tab_entry.prefix_code.unwrap() as usize > CONTROL_CODE);
+    }
+
+    if Some(prefix_code) == queue.next() {
+        /* The prefix code hasn't been added yet, but we were just
+        about to: the KwKwK case. Add the previous string extended
+        with its first byte. */
+        codetab[prefix_code as usize] = Codetab {
+            prefix_code: Some(prev_code),
+            ext_byte: first_byte,
+            len: codetab[prev_code as usize].len + 1,
+            last_dst_pos: codetab[prev_code as usize].last_dst_pos,
+        };
+        dst.push_back(first_byte);
+    } else if codetab[prefix_code as usize].prefix_code.is_none() {
+        // The prefix code is still invalid.
+        return Err(io::Error::new(
+            io::ErrorKind::InvalidData,
+            "invalid prefix code",
+        ));
+    }
+
+    // Output the prefix string, then the extension byte.
+    let len = codetab[prefix_code as usize].len as usize + 1;
+    let last_dst_pos = dst.len();
+    let ct = &codetab[prefix_code as usize];
+    for i in ct.last_dst_pos..ct.last_dst_pos + ct.len as usize {
+        dst.push_back(dst[i]);
+    }
+    dst.push_back(codetab[code as usize].ext_byte);
+    let res = dst[ct.last_dst_pos];
+
+    // Update the code table now that the string has a length and pos.
+    debug_assert!(prev_code != code);
+    codetab[code as usize].len = len as u16;
+    codetab[code as usize].last_dst_pos = last_dst_pos;
+
+    Ok(res)
+}
+
+fn hwunshrink(src: &[u8], uncompressed_size: usize, dst: &mut VecDeque<u8>) -> io::Result<()> {
+    let mut codetab = Codetab::create_new();
+    let mut queue = Box::new(CodeQueue::new());
+    let mut is = BitReader::endian(src, LittleEndian);
+
+    let mut code_size = MIN_CODE_SIZE;
+
+    // Handle the first code separately since there is no previous code.
+    let Ok(Some(curr_code)) = read_code(&mut is, &mut code_size, &mut codetab, &mut queue) else {
```

Comment:
Shouldn't this pass through any error it receives? (Use the `?` operator to do that.)